### PR TITLE
poetryPlugins.poetry-plugin-up: init at 0.2.1

### PR DESCRIPTION
**Pull request by dotlambda**
###### Description of changes
All of these should now work:
```bash
nix-build -A poetry
nix-build -A poetry.plugins.poetry-plugin-up
nix-build -E "with import ./. { }; poetry.withPlugins (ps: with ps; [ poetry-plugin-up ])"
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

closes https://github.com/NixOS/nixpkgs/pull/211231

### CODE REVIEW

File: pkgs/tools/package-management/poetry/default.nix [hunk 1]
```diff
@@ -16,120 +15,28 @@ let
...
-    description = "Python dependency management and packaging made easy";
-    license = licenses.mit;
-    maintainers = with maintainers; [ jakewaksbaum dotlambda ];
-  };
-}
+  # pluginSelector is a function mapping pythonPackages to a list of plugins
+  withPlugins = pluginSelector: python.pkgs.toPythonApplication (
+  # e.g. poetry.withPlugins (ps: with ps; [ poetry-plugin-up ])
+    python.pkgs.poetry.overridePythonAttrs (old: {
+      propagatedBuildInputs = old.propagatedBuildInputs ++ pluginSelector python.pkgs;
```
Comment (dotlambda) [inline pkgs/tools/package-management/poetry/default.nix:29]: Should this be
```suggestion
      propagatedBuildInputs = old.propagatedBuildInputs ++ pluginSelector plugins;
```
?

Comment (K900) [reply pkgs/tools/package-management/poetry/default.nix:29]: Seems reasonable.

**Comment by dotlambda** [comment]
How would I disable tests without disabling them for the poetry package built by Hydra?

File: pkgs/tools/package-management/poetry/default.nix [hunk 2]
```diff
@@ -16,120 +15,28 @@ let
...
+
+      # Propagating dependencies leaks them through $PYTHONPATH which causes issues
+      # when used in nix-shell.
+      postFixup = ''
+        rm $out/nix-support/propagated-build-inputs
+      '';
+
+      passthru = {
+        inherit plugins withPlugins;
```
Comment (dotlambda) [inline pkgs/tools/package-management/poetry/default.nix:38]: What about adding
```suggestion
        inherit plugins withPlugins withAllPlugins;
```
and adding a corresponding expression to `all-packages.nix`. That's an easy way to ensure the plugins are built by hydra.

File: pkgs/tools/package-management/poetry/default.nix [hunk 3]
```diff
@@ -16,120 +15,28 @@ let
...
+      };
+    })
+  );
+in withPlugins (ps: [ ])
```
[the remark this answers is not in the document]
Comment (dotlambda) [reply pkgs/tools/package-management/poetry/default.nix:51]: I just set `doCheck` to true iff zero plugins are selected.

**Comment by dotlambda** [comment]
I'm confused. Running
```
nix-build -E "with import ./. { }; poetry.withPlugins (ps: with ps; [ poetry-audit-plugin ])"
```
produces nothing in `$out/bin`.

**Comment by dotlambda** [comment]
I put `poetry` in the plugin's `propagatedBuildInputs`. Moving to `buildInputs` fixes the issue.
Now it works flawlessly:
```
$ poetry audit -C pkgs/tools/filesystems/rmfuse
# poetry audit report
Loading...
Scanning 26 packages...

  • lxml       installed 4.9.0  affected <4.9.1  CVE CVE-2022-2309
  • pillow     installed 9.1.1  affected <9.2.0  CVE CVE-2022-45198
  • pillow     installed 9.1.1  affected <9.3.0  CVE CVE-2022-45199
  • reportlab  installed 3.6.9  affected >=0     CVE CVE-2020-28463

4 vulnerabilities found in 3 packages
```

File: pkgs/tools/package-management/poetry/default.nix (hunk 2 above)
Comment (dotlambda) [reply pkgs/tools/package-management/poetry/default.nix]: I added `poetry-with-plugins` to `all-packages.nix`.

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 213914` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>poetry</li>
    <li>poetryPlugins.poetry-audit-plugin</li>
    <li>poetryPlugins.poetry-plugin-up</li>
  </ul>
</details>

File: pkgs/tools/package-management/poetry/default.nix (hunk 2 above)
Comment (dotlambda) [reply pkgs/tools/package-management/poetry/default.nix]: Decided to with a top-level attribute `poetryPlugins` instead. That works nicely with nixpkgs-review.

**Pull request merged**